File: 2026/04-April/04-Saturday/Solution.c
#include <stdlib.h>
#include <string.h>

char* decodeCiphertext(char* encodedText, int rows) {
    if (rows == 1) return encodedText;

    int n = strlen(encodedText);
    int cols = n / rows;

    char** matrix = (char**)malloc(rows * sizeof(char*));
    for (int i = 0; i < rows; i++) {
        matrix[i] = (char*)malloc(cols * sizeof(char));
    }

    int idx = 0;
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            matrix[i][j] = encodedText[idx++];
        }
    }

    char* res = (char*)malloc((n + 1) * sizeof(char));
    int k = 0;

    for (int start = 0; start < cols; start++) {
        int i = 0, j = start;
        while (i < rows && j < cols) {
            res[k++] = matrix[i][j];
            i++;
            j++;
        }
    }

    while (k > 0 && res[k - 1] == ' ') k--;
    res[k] = '\0';

    return res;
}

File: 2026/04-April/04-Saturday/solution.py
#updated
class Solution:
    def decodeCiphertext(self, encodedText, rows):
        if rows == 1:
            return encodedText
        
        n = len(encodedText)
        cols = n // rows
        
        matrix = []
        index = 0
        for i in range(rows):
            matrix.append(list(encodedText[index:index + cols]))
            index += cols
        
        res = []
        for start in range(cols):
            i, j = 0, start
            while i < rows and j < cols:
                res.append(matrix[i][j])
                i += 1
                j += 1
        
        return "".join(res).rstrip()
